programs/cleos/main.cpp

```diff
@@ -2957,6 +2961,201 @@ int main( int argc, char** argv ) {
+               std::cout << fc::json::to_pretty_string(v) << std::endl;
+            } else if (type == "transaction") {
+               transaction t;
+               from_variant(unpacked_json, t);
+               auto v = fc::raw::pack<transaction>(t);
+               std::cout << fc::json::to_pretty_string(v) << std::endl;
+            } else if (type == "abi_def") {
+               abi_def t;
+               from_variant(unpacked_json, t);
+               auto v = fc::raw::pack<abi_def>(t);
+               std::cout << fc::json::to_pretty_string(v) << std::endl;
+            } else if (type == "action") {
+               action t;
+               from_variant(unpacked_json, t);
+               auto v = fc::raw::pack<action>(t);
+               std::cout << fc::json::to_pretty_string(v) << std::endl;
+            }
+            success = true;
+
+         } catch (...) {}
+      }
+
+      if (!success) {
+         std::cerr << "ERROR: Unable to convert the JSON";
+         if (!type.empty()) {
+            std::cerr << " to type: " << type;
+         }
+         std::cerr << std::endl;
+      }
+   });
+
+   // unpack hex
+   string packed_hex;
+   auto unpack_hex = convert->add_subcommand("unpack_hex", localized("From packed HEX to JSON form"));
+   unpack_hex->add_option("hex", packed_hex, localized("The packed HEX of built-in types including: signed_block, transaction/action_trace, transaction, action, abi_def"))->required();
+   unpack_hex->add_option("--type", type, (localized("Type of the HEX data, if not specified then some common types are attempted")));
+   unpack_hex->add_option("--abi-file", abi_file, localized("The abi file that contains --type for unpacking"));
+   unpack_hex->callback([&] {
+      EOS_ASSERT( packed_hex.size() >= 2, misc_exception, "No packed HEX data found" );
+      vector<char> packed_blob(packed_hex.size()/2);
+      fc::from_hex(packed_hex, packed_blob.data(), packed_blob.size());
+      bool success = false;
+      if (!type.empty()) {
+         std::optional<abi_def> abi;
+         if (!abi_file.empty()) {
+            abi = json::from_file(abi_file).as<abi_def>();
+         }
+         try {
+            abi_serializer s = abi ? abi_serializer(*abi, abi_serializer::create_yield_function( abi_serializer_max_time )) : abi_serializer{};
+            auto v = s.binary_to_variant(type, packed_blob, fc::microseconds::maximum());
+            std::cout << fc::json::to_pretty_string(v) << std::endl;
+            success = true;
+         } catch (...) {}
+      } else {
+         EOS_ASSERT( abi_file.empty(), misc_exception, "--type required if --abi-file specified");
+      }
+      if (type.empty() || type == "block_state_legacy") {
+         try {
+            auto v = fc::raw::unpack<block_state_legacy>( packed_blob );
+            std::cout << fc::json::to_pretty_string(v) << std::endl;
+            success = true;
+         } catch (...) {}
+      }
+      if (type.empty() || type == "signed_block") {
```

Comment:
```suggestion
      if (!success && (type.empty() || type == "signed_block")) {
```

otherwise when `type.empty() ` we will try all `unpacks` even if the first one succeeds.